packages/react-native/Libraries/AppDelegate/RCTRootViewFactory.mm

```diff
@@ -264,13 +258,15 @@ - (RCTHost *)createReactHost:(NSDictionary *)launchOptions
 
 - (std::shared_ptr<facebook::react::JSRuntimeFactory>)createJSRuntimeFactory
 {
-#if USE_HERMES
-  return std::make_shared<facebook::react::RCTHermesInstance>(nullptr, /* allocInOldGenBeforeTTI */ false);
-#elif USE_THIRD_PARTY_JSC != 1
-  return std::make_shared<facebook::react::RCTJscInstance>();
-#else
-  throw std::runtime_error("No JSRuntimeFactory specified.");
-#endif
+  if (_configuration.jsRuntimeConfiguratorDelegate == nil) {
+    [NSException raise:@"RCTReactNativeFactoryDelegate::createJSRuntimeFactory not implemented"
+                 format:@"Delegate must implement a valid createJSRuntimeFactory method"];
+    return nullptr;
+  }
+  
+  auto jsRuntimeFactory = [_configuration.jsRuntimeConfiguratorDelegate createJSRuntimeFactory];
+  
+  return std::shared_ptr<facebook::react::JSRuntimeFactory>(reinterpret_cast<facebook::react::JSRuntimeFactory *>(jsRuntimeFactory));
```

Comment:
```suggestion
  return std::shared_ptr<facebook::react::JSRuntimeFactory>(reinterpret_cast<facebook::react::JSRuntimeFactory *>(jsRuntimeFactory), &js_runtime_factory_destroy);
```

think more for this. maybe we can just use a custom deleter for the smart pointer
see
```
template< class Y, class Deleter >shared_ptr( Y* ptr, Deleter d ); | (4)
```
from https://en.cppreference.com/w/cpp/memory/shared_ptr/shared_ptr

not sure whether the syntax is correct or should pass into a lambda. can help try this out if needed.

Reply:
I'm not sure if this is needed, I did some tests and destructor is properly called.

I expose destroy function in the api so users can destroy the instance from swift for example when switching engines or something like this

Reply:
it works because we here use `new RCTHermesInstance()`. the default deleter uses `delete`.
i just feel that given we introduced `js_runtime_factory_destroy`, we should use it for shared_ptr deleter. otherwise we have no way to call `js_runtime_factory_destroy`. the shared_ptr is owned within react-native core. 3rd party js runtime like javascriptcore has no way to know when the js runtime shared_ptr will be destroyed.

Reply:
Makes sense, I've passed the destroy function to shared_ptr